pkg/resource/check.go

```diff
@@ -28,33 +31,76 @@ func (cr *CheckRequest) Validate() error {
 func Check(ctx context.Context, request CheckRequest) (*CheckResponse, error) {
 	repo, err := newRepositoryForSource(ctx, request.Source)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "failed to create repository client")
 	}
 
-	// Fetching repository tags
+	// Fetch repository tags
 	allTags, err := registry.Tags(ctx, repo)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to fetch tags")
 	}
 
-	// Sorting tags.
-	var latestTag *semver.Version
-	for _, tag := range allTags {
-		v, err := semver.NewVersion(tag)
+	// Sort tags by semver
+	sortedSemvers := sortBySemver(allTags)
+
+	// chop the list at the index of the requested version, if there was one
+	if request.Version != nil {
+		requestedVersion, err := semver.NewVersion(request.Version.Tag)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, fmt.Sprintf("failed to parse semver in requested version %q", request.Version.Tag))
+		}
+		for i, version := range sortedSemvers {
+			if version.GreaterThanEqual(requestedVersion) {
+				sortedSemvers = sortedSemvers[i:]
+				break
+			}
 		}
-		if latestTag == nil || v.GreaterThan(latestTag) {
-			latestTag = v
+	} else {
+		// if no version was requested, return the latest 10 versions
+		startIndex := len(sortedSemvers) - 10
+		if startIndex < 0 {
+			startIndex = 0
 		}
+		sortedSemvers = sortedSemvers[startIndex:]
 	}
-	if latestTag == nil {
+
+	if len(sortedSemvers) == 0 {
 		return nil, fmt.Errorf("no latest tag found for source %s", request.Source.String())
 	}
 
-	digest, err := getDigestForTag(ctx, repo, latestTag.String())
+	resolvedVersions, err := resolveImageDigests(ctx, sortedSemvers, repo)
 	if err != nil {
 		return nil, err
 	}
-	return &CheckResponse{{Tag: latestTag.String(), Digest: digest}}, nil
+	return &resolvedVersions, nil
+}
+
+func sortBySemver(allTags []string) []semver.Version {
+	allVersions := make([]semver.Version, len(allTags))
+	var j = 0
+	for _, tag := range allTags {
+		v, err := semver.NewVersion(tag)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "skipping tag %q because it does not look like a semver\n", tag)
+		} else {
+			allVersions[j] = *v
+			j++
+		}
+	}
```

Comment:
You can simplify the code by removing the `else` block and using `continue`

```suggestion
	for _, tag := range allTags {
		v, err := semver.NewVersion(tag)
		if err != nil {
			fmt.Fprintf(os.Stderr, "skipping tag %q because it does not look like a semver\n", tag)
			continue
		}
		allVersions[j] = *v
		j++
	}
```